### PR TITLE
feat: Add setup script which also handles dependencies

### PR DESCRIPTION
**Pull request by andiwand**
Currently I use a bunch of setup scripts for my daily developments and I wondered if those can be generated by CMake. In the best case this could be the standard way to setup a dev environment once the dependencies+built is figured out

Sourcing our python scripts dir is currently necessary to find the ODD directory. This can later be generalized by an environment variable.

Mid term we could also think about adding a dependency build script to the repo and have a CI job monitoring the dev environment setup for one or more systems

### CODE REVIEW

File: cmake/setup_withdeps.sh.in [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+
+# make ACTS binaries available
+export PATH="$script_dir/bin:${PATH}"
+export LD_LIBRARY_PATH="$script_dir/lib:${LD_LIBRARY_PATH}"
+export DYLD_LIBRARY_PATH="$script_dir/lib:${DYLD_LIBRARY_PATH}"
+
+# activate dependencies
+. @ROOT_DIR@/../bin/thisroot.sh
+. @Geant4_DIR@/../../../bin/geant4.sh
+. @DD4hep_DIR@/bin/thisdd4hep.sh
```
Comment (paulgessinger) [inline cmake/setup_withdeps.sh.in:40]: Could we include podio and edm4hep conditionally as well?

File: cmake/setup_withdeps.sh.in [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+# modify python environment to include some ACTS helpers
+export PYTHONPATH="$ACTS_SOURCE_DIR/Examples/Scripts/Python:${PYTHONPATH}"
+
+# make ACTS binaries available
+export PATH="$script_dir/bin:${PATH}"
+export LD_LIBRARY_PATH="$script_dir/lib:${LD_LIBRARY_PATH}"
+export DYLD_LIBRARY_PATH="$script_dir/lib:${DYLD_LIBRARY_PATH}"
+
+# activate dependencies
+. @ROOT_DIR@/../bin/thisroot.sh
```
Comment (paulgessinger) [inline cmake/setup_withdeps.sh.in:38]: This should probably be conditional on whether each dependency is built.

File: cmake/setup_withdeps.sh.in [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+fi
+if [[ -d "@DD4hep_DIR@" ]]; then
+  . @DD4hep_DIR@/bin/thisdd4hep.sh
+fi
+if [[ -d "@podio_DIR@" ]]; then
+  export ROOT_INCLUDE_PATH=@podio_DIR@/include:"${ROOT_INCLUDE_PATH}"
+fi
+if [[ -d "@EDM4HEP_DIR@" ]]; then
+  export ROOT_INCLUDE_PATH=@EDM4HEP_DIR@/include:"${ROOT_INCLUDE_PATH}"
+fi
```
Comment (paulgessinger) [inline cmake/setup_withdeps.sh.in:52]: Can you also add these to the library paths?

**Comment by andiwand** [comment]
It would be nice to monitor this somehow in the CI. Our worker have all dependencies installed system wide. Maybe we can purge the default system library path somehow? Do you see a solution for this @paulgessinger?

File: cmake/setup_withdeps.sh.in [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+if [[ -d "@Geant4_DIR@" ]]; then
+  . @Geant4_DIR@/../../../bin/geant4.sh
+fi
+if [[ -d "@DD4hep_DIR@" ]]; then
+  . @DD4hep_DIR@/bin/thisdd4hep.sh
+fi
+if [[ -d "@podio_DIR@" ]]; then
+  export LD_LIBRARY_PATH="@podio_DIR@/lib:${LD_LIBRARY_PATH}"
+  export DYLD_LIBRARY_PATH="@podio_DIR@/lib:${DYLD_LIBRARY_PATH}"
+  export ROOT_INCLUDE_PATH="@podio_DIR@/include:${ROOT_INCLUDE_PATH}"
```
Comment (andiwand) [inline cmake/setup_withdeps.sh.in:50]:
```suggestion
  export ROOT_INCLUDE_PATH="@podio_DIR@/include:${ROOT_INCLUDE_PATH}"
  export PYTHONPATH="@podio_DIR@/python:${PYTHONPATH}"
```

**Comment by andiwand** [comment]
podio is still failing

File: .github/workflows/builds.yml [hunk 5]
```diff
@@ -146,12 +146,7 @@ jobs:
           PYTEST_MD_REPORT_OUTPUT: pytest.md
         run: >
           /usr/local/bin/geant4-config --install-datasets
-          && source /usr/local/bin/thisroot.sh
-          && source /usr/local/bin/thisdd4hep_only.sh
-          && source /usr/local/bin/geant4.sh
-          && source build/python/setup.sh
-          && export PYTHONPATH=/usr/local/python:$PYTHONPATH
-          && export LD_LIBRARY_PATH=$PWD/build/thirdparty/OpenDataDetector/factory:$LD_LIBRARY_PATH
+          && source build/this_acts_withdeps.sh
```
Comment (andiwand) [inline .github/workflows/builds.yml:149]:
```suggestion
          && source build/this_acts_withdeps.sh
          && echo "HIHIHI"
          && cat build/this_acts_withdeps.sh
          && echo &PYTHONPATH
```

File: .github/workflows/builds.yml [hunk 6]
```diff
@@ -146,12 +146,10 @@ jobs:
...
+          && echo "HIHIHI"
+          && cat build/this_acts_withdeps.sh
+          && echo &PYTHONPATH
```
Comment (andiwand) [inline .github/workflows/builds.yml:152]:
```suggestion
          && echo $PYTHONPATH
```

File: .github/workflows/builds.yml [hunk 7]
```diff
@@ -146,12 +146,10 @@ jobs:
...
-          && source /usr/local/bin/thisroot.sh
-          && source /usr/local/bin/thisdd4hep_only.sh
-          && source /usr/local/bin/geant4.sh
-          && source build/python/setup.sh
-          && export PYTHONPATH=/usr/local/python:$PYTHONPATH
-          && export LD_LIBRARY_PATH=$PWD/build/thirdparty/OpenDataDetector/factory:$LD_LIBRARY_PATH
+          && source build/this_acts_withdeps.sh
+          && echo "HIHIHI"
+          && cat build/this_acts_withdeps.sh
+          && echo $PYTHONPATH
```
Comment (andiwand) [inline .github/workflows/builds.yml:152]:
```suggestion
```

File: cmake/setup_withdeps.sh.in [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+export PATH="$script_dir/bin:${PATH}"
+export LD_LIBRARY_PATH="$script_dir/lib:${LD_LIBRARY_PATH}"
+export DYLD_LIBRARY_PATH="$script_dir/lib:${DYLD_LIBRARY_PATH}"
+
+# activate dependencies if present
+if [[ -d "@ROOT_DIR@" ]]; then
+  . @ROOT_BINDIR@/thisroot.sh
+fi
+if [[ -d "@Geant4_DIR@" ]]; then
+  . @Geant4_INCLUDE_DIR@/../bin/geant4.sh
```
Comment (andiwand) [inline cmake/setup_withdeps.sh.in:42]:
```suggestion
  . @Geant4_INCLUDE_DIR@/../../bin/geant4.sh
```

**Review by paulgessinger — APPROVED** [review]

**Comment by acts-project-service** [comment]
## :red_circle: Athena integration test results [6a2554575bced122e145edeab27ae2e4ed1c4200]
### Build job with this PR failed!
Please investigate the [build job](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/36524968) for the [pipeline](https://gitlab.cern.ch/acts/acts-athena-ci/-/pipelines/6968567)!

**Pull request merged**